src/liteclient.cc

```diff
@@ -205,6 +207,117 @@ boost::container::flat_map<int, Uptane::Target> LiteClient::getTargets() {
+  data::ResultCode::Numeric update_result = data::ResultCode::Numeric::kUnknown;
+  std::string download_reason;
+  auto current_target = getCurrent();
+
+  Uptane::Target target_to_update{target};
+
+  UpdateType update_type = UpdateType::kNewTargetUpdate;
+  if (force_update) {
+    update_type = UpdateType::kTargetForcedUpdate;
+  } else if (target.filename() == current_target.filename() || !isTargetValid(target)) {
+    update_type = UpdateType::kCurrentTargetSync;
+    target_to_update = current_target;
+  }
+
+  // set an update correlation ID to "bind" all report events generated during a single update together
+  // so the backend can know to which of the updates specific events are related to
+  Target::setCorrelationID(target_to_update);
+
+  // prepare for an update
+  switch (update_type) {
+    case UpdateType::kNewTargetUpdate: {
+      logTarget("Updating Active Target: ", current_target);
+      logTarget("To New Target: ", target_to_update);
+      download_reason = "Updating from " + current_target.filename() + " to " + target_to_update.filename();
+
+      appsInSync();
+      break;
+    }
+    case UpdateType::kCurrentTargetSync: {
+      logTarget("Syncing current Target: ", target_to_update);
+      download_reason = "Syncing current Target " + target_to_update.filename();
+
+      if (appsInSync() && target_to_update.sha256Hash() == current_target.sha256Hash()) {
+        setAppsNotChecked();
+        return data::ResultCode::Numeric::kAlreadyProcessed;
+      }
+
+      break;
+    }
+    case UpdateType::kTargetForcedUpdate: {
+      logTarget("Updating to Target: ", target);
+      download_reason = "Manual update to Target " + target.filename();
+      break;
+    }
+    default: {
+      // this shouldn't really happen at the runtime
+      throw std::runtime_error("Unsupported type of update:  " + std::to_string(update_type));
+    }
+  }
+
+  // do update, i.e. download and install
+  do {
+    update_result = download(target_to_update, download_reason);
+
+    // TODO: it should be really just `download` and `install`
+    if (update_result != data::ResultCode::Numeric::kOk) {
+      break;
+    }
+
+    if (VerifyTarget(target_to_update) != TargetStatus::kGood) {
+      // why do we do it here not inside of packman->download() ???
+      data::InstallationResult res{data::ResultCode::Numeric::kVerificationFailed, "Downloaded target is invalid"};
+      // why do we send notifyInstallFinished if notifyInstallStarted hasn't been called ?
+      notifyInstallFinished(target_to_update, res);
+      LOG_ERROR << "Downloaded target is invalid";
+      update_result = res.result_code.num_code;
+      break;
+    }
+
+    update_result = install(target_to_update);
+
+    if (update_result == data::ResultCode::Numeric::kOk) {
+      current_target_ = target_to_update;
+      http_client->updateHeader("x-ats-target", current_target_.filename());
+    }
+  } while (false);  // update scope
```

Comment:
Yes, that's my plan, I put the code under do {} while (false) because I saw it as a candidate for dedicated method, this is just WIP and I didn't want to do it until I see the full picture.

Reply:
interesting. I love seeing techniques people use while developing code